cloud/network-load-balancer/main.tf

```diff
@@ -0,0 +1,227 @@
+# External Network Load Balancer Module
+#
+# This module creates a regional external Network Load Balancer on GCP for exposing
+# services to the internet. It supports both TCP and UDP protocols (required for WireGuard).
+#
+# Example Usage (WireGuard VPN Gateway):
+#
+#   module "wireguard_nlb" {
+#     source = "./cloud/network-load-balancer"
+#
+#     name   = "wireguard-gateway"
+#     region = "us-central1"
+#
+#     protocols = ["TCP", "UDP"]
+#     port_range = {
+#       start = 51820
+#       end   = 51820
+#     }
+#
+#     instance_groups = [{
+#       instance_group = google_compute_instance_group.wireguard.id
+#       balancing_mode = "CONNECTION"
+#     }]
+#
+#     health_check = {
+#       protocol           = "TCP"
+#       port               = 51820
+#       check_interval_sec = 10
+#       timeout_sec        = 5
+#     }
+#   }
+#
+# Example Usage (Static External IP):
+#
+#   resource "google_compute_address" "wireguard_ip" {
+#     name         = "wireguard-ip"
+#     region       = "us-central1"
+#     address_type = "EXTERNAL"
+#   }
+#
+#   module "wireguard_nlb" {
+#     source = "./cloud/network-load-balancer"
+#
+#     name                = "wireguard-gateway"
+#     region              = "us-central1"
+#     external_ip_address = google_compute_address.wireguard_ip.name
+#
+#     protocols = ["TCP", "UDP"]
+#     port_range = {
+#       start = 51820
+#       end   = 51820
+#     }
+#
+#     instance_groups = [{
+#       instance_group = google_compute_instance_group.wireguard.id
+#       balancing_mode = "CONNECTION"
+#     }]
+#   }
+#
+# Example Usage (HTTP Health Check):
+#
+#   module "http_nlb" {
+#     source = "./cloud/network-load-balancer"
+#
+#     name   = "http-service"
+#     region = "us-central1"
+#
+#     protocols = ["TCP"]
+#     port_range = {
+#       start = 80
+#       end   = 80
+#     }
+#
+#     instance_groups = [{
+#       instance_group = google_compute_instance_group.http_servers.id
+#       balancing_mode = "UTILIZATION"
+#       max_utilization = 0.8
+#     }]
+#
+#     health_check = {
+#       protocol           = "HTTP"
+#       port               = 8080
+#       request_path       = "/health"
+#       check_interval_sec = 10
+#       timeout_sec        = 5
+#     }
+#   }
+
+terraform {
+  required_providers {
+    google = {
+      source  = "hashicorp/google"
+      version = "7.11.0"
+    }
+  }
+}
+
+# Optional: Create or use existing external IP address
+resource "google_compute_address" "external_ip" {
```

Comment:
The ip address should be required to be provided so that users can destroy and reapply the nlb constantly without allocating a new ip which could break upstream dns records